src/components/MAPI/organizations/AccessControl/AccessControlRoleSubjects.tsx

```diff
@@ -0,0 +1,377 @@
+import { Box, Text } from 'grommet';
+import { FlashMessage, messageTTL, messageType } from 'lib/flashMessage';
+import PropTypes from 'prop-types';
+import React, { useEffect, useReducer } from 'react';
+import AccessControlSubjectSet, {
+  IAccessControlSubjectSetItem,
+} from 'UI/Display/MAPI/AccessControl/AccessControlSubjectSet';
+import AccessControlSubjectSetItem from 'UI/Display/MAPI/AccessControl/AccessControlSubjectSetItem';
+
+import {
+  AccessControlSubjectTypes,
+  IAccessControlRoleItem,
+  IAccessControlRoleSubjectItem,
+} from '../../../UI/Display/MAPI/AccessControl/types';
+import { getUserNameParts } from './utils';
+
+interface IStateValue {
+  isAdding: boolean;
+  isLoading: boolean;
+  namesLoading: string[];
+}
+
+type State = Record<AccessControlSubjectTypes, IStateValue>;
+
+interface IAction {
+  type: 'startAdding' | 'stopAdding' | 'startLoading' | 'stopLoading' | 'reset';
+  subjectType: AccessControlSubjectTypes;
+  subjectName?: string;
+}
+
+const initialState: State = {
+  [AccessControlSubjectTypes.Group]: {
+    isAdding: false,
+    isLoading: false,
+    namesLoading: [],
+  },
+  [AccessControlSubjectTypes.User]: {
+    isAdding: false,
+    isLoading: false,
+    namesLoading: [],
+  },
+  [AccessControlSubjectTypes.ServiceAccount]: {
+    isAdding: false,
+    isLoading: false,
+    namesLoading: [],
+  },
+};
+
+const reducer: React.Reducer<State, IAction> = (state, action) => {
+  const currSubject = state[action.subjectType];
+
+  switch (action.type) {
```

Comment:
Could potentially switch to using `immer`'s `produce()` function here, at some point